### PR TITLE
[Core][Bugfix] Support prompt_logprobs returned with speculative decoding

### LINKED ISSUE #7742: [Bug]: Requesting Prompt Logprobs with an MLP Speculator Crashes the Server
### Your current environment

Using the latest vLLM off of `main`.

### 🐛 Describe the bug

When running the online server with a model with an MLP speculator, sending a request that request prompt logprobs causes the server to crash with an `AssertionError`.

Stacktrace:
```
Traceback (most recent call last):
  File "/workspace/my-vllm/lib64/python3.11/site-packages/vllm/entrypoints/openai/rpc/server.py", line 125, in generate
    async for request_output in results_generator:
  File "/workspace/my-vllm/lib64/python3.11/site-packages/vllm/engine/async_llm_engine.py", line 1054, in generate
    async for output in await self.add_request(
  File "/workspace/my-vllm/lib64/python3.11/site-packages/vllm/engine/async_llm_engine.py", line 114, in generator
    raise result
  File "/workspace/my-vllm/lib64/python3.11/site-packages/vllm/engine/async_llm_engine.py", line 55, in _log_task_completion
    return_value = task.result()
                   ^^^^^^^^^^^^^
  File "/workspace/my-vllm/lib64/python3.11/site-packages/vllm/engine/async_llm_engine.py", line 920, in run_engine_loop
    result = task.result()
             ^^^^^^^^^^^^^
  File "/workspace/my-vllm/lib64/python3.11/site-packages/vllm/engine/async_llm_engine.py", line 863, in engine_step
    request_outputs = await self.engine.step_async(virtual_engine)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/my-vllm/lib64/python3.11/site-packages/vllm/engine/async_llm_engine.py", line 332, in step_async
    output = await self.model_executor.execute_model_async(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/my-vllm/lib64/python3.11/site-packages/vllm/executor/gpu_executor.py", line 170, in execute_model_async
    output = await make_async(self.driver_worker.execute_model
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/my-vllm/lib64/python3.11/site-packages/torch/utils/_contextlib.py", line 116, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/my-vllm/lib64/python3.11/site-packages/vllm/spec_decode/spec_decode_worker.py", line 387, in execute_model
    return self._run_no_spec(execute_model_req,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/contextlib.py", line 81, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/workspace/my-vllm/lib64/python3.11/site-packages/vllm/spec_decode/spec_decode_worker.py", line 481, in _run_no_spec
    self.previous_hidden_states.update(
  File "/workspace/my-vllm/lib64/python3.11/site-packages/vllm/sequence.py", line 1199, in update
    assert len(seq_group_metadata_list) == len(hidden_states)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError
```

### To Reproduce

Run a server with an MLP speculator, eg. one of IBM's granite models:
```
vllm serve ibm-granite/granite-3b-code-instruct --speculative-model ibm-granite/granite-3b-code-instruct-accelerator --use-v2-block-manager --enforce-eager
```

Send an `echo` request with logprobs requested for the prompt tokens:
```
curl http://localhost:8000/v1/completions \
  -H "Content-Type: application/json" \
  -d '{
      "model": "ibm-granite/granite-3b-code-instruct",
      "prompt": "Hello World",
      "echo": 1,
      "logprobs": 1,
      "temperature": 0
  }'
```

### PR DESCRIPTION
**Pull request by tjohnson31415**
In the current code, requesting prompt_logprobs while using speculative decoding can results in crashes. With an MLPSpeculator, an AssertionError is triggered (https://github.com/vllm-project/vllm/issues/7742). With an LLM-based draft model, the behavior has some randomness due to uninitialized data and can "work" (but first speculated token is random), result in a `RuntimeError: CUDA error: device-side assert triggered`, or a conversion error like
```
  File "/workspace/my-vllm/lib64/python3.11/site-packages/vllm/engine/async_llm_engine.py", line 356, in step_async
    request_outputs = self._process_model_outputs(
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/my-vllm/lib64/python3.11/site-packages/vllm/engine/llm_engine.py", line 1232, in _process_model_outputs
    self.output_processor.process_outputs(seq_group, outputs)
  File "/workspace/my-vllm/lib64/python3.11/site-packages/vllm/engine/output_processor/multi_step.py", line 89, in process_outputs
    self._process_seq_outputs(seq, valid_samples,
  File "/workspace/my-vllm/lib64/python3.11/site-packages/vllm/engine/output_processor/multi_step.py", line 130, in _process_seq_outputs
    new_char_count = self.detokenizer.decode_sequence_inplace(
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/my-vllm/lib64/python3.11/site-packages/vllm/transformers_utils/detokenizer.py", line 124, in decode_sequence_inplace
    read_offset) = detokenize_incrementally(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/my-vllm/lib64/python3.11/site-packages/vllm/transformers_utils/detokenizer.py", line 289, in detokenize_incrementally
    new_tokens = tokenizer.convert_ids_to_tokens(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspace/my-vllm/lib64/python3.11/site-packages/transformers/tokenization_utils_fast.py", line 412, in convert_ids_to_tokens
    tokens.append(self._tokenizer.id_to_token(index))
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
OverflowError: out of range integral type conversion attempted
```

The fixes in this PR are to ensure that the speculative decoding code handles the extra entries in the model outputs from the prompt tokens instead of crashing. By fixing these crashes, prompt_logprobs can now be returned when using spec decoding.

Other changes with this PR:
- refactor `INVALID_TOKEN_ID`, renaming it to `VLLM_INVALID_TOKEN_ID` and moving it to sampler.py (avoids a circular import error that arises from the next bullet point)
- use `VLLM_INVALID_TOKEN_ID` in place of "magic number" -1 for invalid token ids
- remove `run_logprob_correctness_test` in spec_decode tests to remove some code duplication; just check logprobs in `run_equality_correctness_test` if they are requested

FIX https://github.com/vllm-project/vllm/issues/7742

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/sequence.py [hunk 1]
```diff
@@ -1125,6 +1127,17 @@ def __eq__(self, other: object):
                           self.__class__) and self.outputs == other.outputs
 
 
+def get_all_seq_data_entries(
+    seq_group_metadata_list: List[SequenceGroupMetadata]
+) -> List[Tuple[int, SequenceData]]:
+    """Given a list of SequenceGroupMetadata, create a dict of
```
Comment (njhill) [inline vllm/sequence.py:1133]:
```suggestion
    """Given a list of SequenceGroupMetadata, create a list of tuples of
```

File: vllm/transformers_utils/detokenizer.py [hunk 2]
```diff
@@ -282,14 +280,14 @@ def detokenize_incrementally(
     assert prev_tokens is not None
 
     # If the new token id is out of bounds, return an empty string.
-    if new_token_id >= len(tokenizer):
-        new_tokens = [""]
-    else:
+    if 0 < new_token_id < len(tokenizer):
```
Comment (njhill) [inline vllm/transformers_utils/detokenizer.py:283]: should this be
```suggestion
    if 0 <= new_token_id < len(tokenizer):
```
?

File: vllm/spec_decode/spec_decode_worker.py [hunk 3]
```diff
@@ -451,14 +454,43 @@ def _serialize_sampler_output_no_logprobs(
...
+            `CompletionSequenceGroupOutput` objects with only token IDs
+            populated.
         """
-        seq_ids = get_all_seq_ids(execute_model_req.seq_group_metadata_list)
-        sampled_token_ids_list = sampler_output.sampled_token_ids.tolist()
+        seq_output_prompt_logprobs = [
+            seq.is_prompt and seq.sampling_params.prompt_logprobs is not None
+            and seq.sampling_params.prompt_logprobs > 0
+            for seq in execute_model_req.seq_group_metadata_list
+        ]
```
Comment (njhill) [inline vllm/spec_decode/spec_decode_worker.py:464]: This could be a generator too

**Review by njhill — COMMENTED** [review]
Thanks @tjohnson31415 looks great, just a couple of minor comments

File: vllm/spec_decode/spec_decode_worker.py (hunk 3 above)
Comment (tjohnson31415) [reply vllm/spec_decode/spec_decode_worker.py:463]: This one is iterated twice in an `any()` and a `zip()`, so I think just making it a list is better.

**Review by njhill — APPROVED** [review]
Thanks @tjohnson31415 for all of the hard work on this!

**Comment by njhill** [comment]
cc @afeldman-nm @comaniac @SolitaryThinker @LiuXiaoxuanPKU in case one of you wants to take a look at this before it's merged

**Review by mgoin — APPROVED** [review]
`_serialize_sampler_output_no_logprobs` seems like it could be pretty slow so it could be nice to have a short-circuit if there are no sequences that require prompt logprobs.

Otherwise this looks good to me!

**Comment by njhill** [comment]
Thanks @mgoin! The flow in the no-prompt-logprobs case actually looks pretty close to what's there before.

**Pull request merged**